aten/src/ATen/native/cuda/AdaptiveAveragePooling.cu

```diff
@@ -220,46 +438,115 @@ namespace {
         "empty");
     }
 
-    TORCH_CHECK((input.ndimension() == 3 || input.ndimension() == 4),
-      "non-empty 3D or 4D (batch mode) tensor expected for input");
     Tensor input_ = input;
-    int64_t grid_x = input.size(-3);
-    if (input.ndimension() == 4) {
-       input_ = input.contiguous();
-       grid_x *= input_.size(-4);
-    }
-    int64_t sizeD  = input_.size(-3);
-    int64_t isizeH = input_.size(-2);
-    int64_t isizeW = input_.size(-1);
-
-    int64_t istrideD = input_.stride(-3);
-    int64_t istrideH = input_.stride(-2);
-    int64_t istrideW = input_.stride(-1);
-
-    int64_t osizeH = output_size[0];
-    int64_t osizeW = output_size[1];
-    if (input.ndimension() == 4) {
-       output.resize_({input_.size(-4), sizeD, osizeH, osizeW});
+    if (input.suggest_memory_format() == at::MemoryFormat::ChannelsLast) {
+      // special case for tensor memory format in channels_last
+      TORCH_CHECK(input.ndimension() == 4,
+        "non-empty 4D (batch mode) tensor expected for input with channels_last layout");
+
```

Comment:
Are we comfortable with non-dense input? default code path tries to make it contiguous (non overlapping+dense)

Reply:
Answering both question on not calling contiguous for input tensor: Let me benchmark them to get some ideas.

My intuition is that: the kernel is basically bandwidth bounded, if we call a contiguous on non-packed input, we are basically doubling memory traffic as well as launch overhead. If our kernel handles arbitrary strides, we should probably just rely on that and not call contiguous explicitly.
I'll benchmark and let you know what I find.

Reply:
Thanks. As for me I just concerned if we support/or not arbitrary strided input.

Reply:
Benchmark shows pretty good speed up. Contiguous call done by TensorIterator looks suspiciously slow. It's noted and I'll take a look later. 

 I'm going to leave the implementation as is and put a unit test for non-contiguous call.